lib/srv/db/sqlserver/kinit/kinit_test.go

```diff
@@ -0,0 +1,164 @@
+	cmd := exec.Command("")
+	cmd.Err = errors.New("bad command")
+	return cmd
+
+}
+
+type testCertGetter struct {
+	pass bool
+}
+
+func (t *testCertGetter) GetCertificateBytes(context.Context) (*WindowsCAAndKeyPair, error) {
+	if t.pass {
+		return &WindowsCAAndKeyPair{}, nil
+	}
+	return nil, errors.New("could not get cert bytes")
+
+}
+
+type testCase struct {
+	name           string
+	initializer    *PKInit
+	expectErr      func(t require.TestingT, err error, msgAndArgs ...interface{})
+	expectCacheNil func(t require.TestingT, object interface{}, msgAndArgs ...interface{})
+}
+
+func step(t *testing.T, name string, cg CommandGenerator, c *testCertGetter, expectErr func(t require.TestingT, err error, msgAndArgs ...interface{}), expectNil func(t require.TestingT, object interface{}, msgAndArgs ...interface{})) *testCase {
+	dir := t.TempDir()
+	var err error
+	dir, err = os.MkdirTemp(dir, "krb5_cache")
+	require.NoError(t, err)
+
+	return &testCase{
+		name: name,
+		initializer: New(NewCommandLineInitializerWithCommand(nil,
+			"alice",
+			"example.com",
+			"host.example.com",
+			"host.example.com",
+			dir, nil, cg, c)),
+		expectErr:      expectErr,
+		expectCacheNil: expectNil,
+	}
+}
+
+func TestNewWithCommandLineProvider(t *testing.T) {
+
+	cases := []*testCase{
+		step(t, "TestCommandSuccessCase", &staticCache{t: t, pass: true}, &testCertGetter{pass: true}, require.NoError, require.NotNil),
+		step(t, "TestCertificateFailureCase", &staticCache{t: t, pass: true}, &testCertGetter{pass: false}, require.Error, require.Nil),
+		step(t, "TestCommandFailureCase", &staticCache{t: t, pass: false}, &testCertGetter{pass: true}, require.Error, require.Nil),
+		// Final case returns an error along with a non nil credentials cache that essentially has all zeroed fields;
+		// &credentials.CCache{Version:0x0, Header:credentials.header{length:0x0, fields:[]credentials.headerField(nil)}, DefaultPrincipal:credentials.principal{Realm:"", PrincipalName:types.PrincipalName{NameType:0, NameString:[]string(nil)}}, Credentials:[]*credentials.Credential(nil), Path:""}
+		step(t, "TestBadCacheData", &badCache{t: t}, &testCertGetter{pass: true}, require.Error, require.NotNil),
+	}
+
+	for _, c := range cases {
+		require.True(t, t.Run(c.name, func(t *testing.T) {
+			cc, err := c.initializer.UseOrCreateCredentialsCache(context.Background())
+			c.expectErr(t, err)
+			c.expectCacheNil(t, cc)
+		}))
+	}
+
+}
+
+const (
+	expectedConfString = `[libdefaults]
+ default_realm = example.com
+ rdns = false
+
+
+[realms]
+ example.com = {
+  kdc = host.example.com
+  admin_server = host.example.com
+  pkinit_eku_checking = kpServerAuth
+  pkinit_kdc_hostname = host.example.com
+ }`
+)
+
+func TestKRBConfString(t *testing.T) {
+	cli := NewCommandLineInitializerWithCommand(nil,
+		"alice",
+		"example.com",
+		"host.example.com",
+		"host.example.com",
+		"",
+		nil, &staticCache{t: t, pass: true}, &testCertGetter{pass: true})
+
+	tmp, err := os.MkdirTemp("", "kinit")
```

Comment:
Could use `t.TempDir()` here instead of `os.MkdirTemp` so you don't need to defer something to tidy up :')